### PR TITLE
feat: get product categories

### PR DESCRIPTION
**Pull request by Amal-Salam**
###  What does the PR do?

     This PR addresses issue #257 by implementing the functionality to get product categories from the database.

### How should this be manually tested?
To test this feature, you can send a GET request to the following endpoint:
- [x] GET /api/v1/products/categories 

```json
    Authorization: Bearer <token>
```
     Successful Response

     On a successful products category retrieval, the API should return a 200  status code.

```json
{
  "status_code": 200,
  "categories": [
    {
      "id": "aee53410-ad20-4ac1-9876-cb650c75b88b",
      "created_at": "2024-07-24T20:08:05.581Z",
      "updated_at": "2024-07-24T20:08:05.581Z",
      "name": "Electronics",
      "description": "Electronic devices and accessories",
      "slug": "electronics",
      "parent_id": null
    }
```
success

Authentication Error:

When the current user is not properly authenticated to get the product category detail:

```json
  {
     "status": "error",
     "message": "User is currently unauthorized, kindly authenticate to continue",
     "status_code": 401
  }
```
no token image

Invalid Query Parameters

On Invalid error, if the query parameters are incorrect:

```json
{
  "error": {
    "code": "INVALID_QUERY_PARAMETER",
    "message": "The provided query parameter is invalid.",
    "details": {
      "invalid_parameter": "limit",
      "reason": "Must be a positive integer"
    }
  }
}
```
invalid query params

Validation:

If any required field is missing or invalid, the response should return a 400 Bad Request status code with appropriate validation error messages.

```json
{
  "message": "Cannot GET /api/v1/products/categorie",
  "error": "Not Found",
  "statusCode": 404
}
```
validation_error_image

Task
 - [x] Create the API endpoint for retrieving product categories.
 - [x] Implement efficient database querying to fetch category data.
 - [x] Include proper indexing on the categories table for optimal performance.
 - [x] Implement optional query parameters for advanced use cases (limit, offset).
 - [x] Ensure proper error handling for invalid requests or server-side issues.
 - [x] Write unit tests for the category retrieval logic.
 - [x] Update API documentation to reflect the new endpoint and its usage.

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/257)

### CODE REVIEW

**Review by rugue — CHANGES_REQUESTED** [review]
pls resolve merge conflicts. pull from dev before making PR

**Comment by Heba-WebDev** [comment]
Wrong branch name, change it please to feat/products-category

File: .vscode/settings.json [hunk 1]
```diff
@@ -10,9 +10,14 @@
     "statusBar.background": "#1a1f15",
     "statusBarItem.hoverBackground": "#333d2a",
     "statusBar.foreground": "#e7e7e7",
```
Comment (davidolanrewaju) [inline .vscode/settings.json:12]: Please remove these settings, line 13-14

File: src/modules/product-category/entities/product-category.entity.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { Entity, PrimaryGeneratedColumn, Column, Index, OneToMany } from 'typeorm';
+import { ApiProperty } from '@nestjs/swagger';
+import { AbstractBaseEntity } from '../../../entities/base.entity';
+import { Product } from '../../products/entities/product.entity';
```
Comment (davidolanrewaju) [inline src/modules/product-category/entities/product-category.entity.ts:4]: Add a an extra line after all imports

File: src/modules/product-category/entities/product-category.entity.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+@Entity('categories')
+export class ProductCategory extends AbstractBaseEntity {
```
Comment (davidolanrewaju) [inline src/modules/product-category/entities/product-category.entity.ts:6]: Extending AbstractBaseEntity doesn't allow for a new id column, please remove it(primary generated column)

File: src/modules/product-category/product-category.controller.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { Controller, Get, Post, Body, Param, Patch, Delete, UseGuards,Query } from '@nestjs/common';
+import { ProductCategoryService } from './product-category.service';
+import { CreateProductCategoryDto } from './dto/create-product-category.dto';
+import { UpdateProductCategoryDto } from './dto/update-product-category.dto';
+import { AuthGuard } from '../../guards/auth.guard';
+import { ApiTags, ApiBody, ApiParam } from '@nestjs/swagger';
+
+@ApiTags('products/categories')
```
Comment (davidolanrewaju) [inline src/modules/product-category/product-category.controller.ts:8]: The auth guard is globally initiated, please remove

File: src/modules/products/products.controller.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+@Controller('products')
+export class ProductController {
+  constructor(private readonly productsService: ProductsService) {}
+
+  @UseGuards(AuthGuard)
+  @ApiTags('Products')
+  @ApiBody({ type: CreateProductDto })
+  @Post()
+  async createProduct(@Body() createProductDto: CreateProductDto, @Req() req): Promise<any> {
+    const product = await this.productsService.create(createProductDto);
```
Comment (davidolanrewaju) [inline src/modules/products/products.controller.ts:16]: Place all db calls in a try catch block

**Comment by Amal-Salam** [comment]
all comments resolved.
thank you

**Pull request closed without merging**